nrf_modem/include/nrf_socket.h

```diff
@@ -17,7 +17,22 @@
 extern "C" {
 #endif
 
+#if !defined(__GNUC__) || (__GNUC__ == 0)
 typedef int32_t ssize_t;
+#else
+#include <sys/types.h>
+#ifdef __SES_ARM
+typedef int32_t ssize_t;
+#endif
+#endif
+
+#ifdef __cplusplus
+#   ifdef __GNUC__
```

Comment:
Should this not be `__GNUG__` ? `__GNUC__` is the C compiler.

Reply:
It's not clear to me how the C compiler can be used to compile C++ code. `g++` is the GNU C++ compiler and `gcc` is the GNU C compiler. They are both part of GCC but they are different compilers. The C++ compiler defines `__GNUG__` and the C compiler defines `__GNUC__`, as far as I understand.

So I think here `__GNUG__` is correct. The fact that it works with `__GNUC__` is perhaps because the `g++` compiler supports `restrict`.

Reply:
Please see: https://gcc.gnu.org/onlinedocs/gcc-3.3/cpp/Common-Predefined-Macros.html
a)
__GNUC__
__GNUC_MINOR__
__GNUC_PATCHLEVEL__
These macros are defined by all GNU compilers that use the C preprocessor: C, C++, and Objective-C. 

b) __GNUG__
The GNU C++ compiler defines this. Testing it is equivalent to testing (__GNUC__ && __cplusplus).

So if we use GNUG we clean up the code, I agree. But in general GNUC is also set by GNU C++ compiler too, meaning the initial commit solution is correct and would explain why it works in our code base. We use c++ in a FreeRTOS environment without Zephyr yet.

Reply:
I personally like making the combination of C++ (outer ifdef) and GNUC more explicit more. Because you see that the explicit keyword is only defined away when using C++ on non GNU compilers.

Reply:
Thanks for helping me look into this; I agree with you that it's a bit clearer with two defines.